### DISCUSSION TITLE
Rendering in background thread

### DISCUSSION THREAD

**Discussion by machur** [pyvista/pyvista#4006] [Q&A]
Hi, I'm working on a stand-alone viewer application that uses PyVista for rendering. It takes about one minute to render my scene and unfortunately the whole Qt window freezes during that process. I've built the rendering window based on the QtInteractor class, but I would like to find a way to render the scene in the background without blocking the main window. Would you please advice?

During my investigation I came across the BackgroundPlotter class and the issues https://github.com/pyvista/pyvista/issues/280 https://github.com/pyvista/pyvista/pull/293 https://github.com/pyvista/pyvista/pull/304, but the solution proposed in these PRs doesn't seem to be working on my setup (Windows, Python 3.8, running Python from command-line).

**Comment by banesullivan** [comment]
Could you provide some sort of minimalish example to reproduce the blocking behavior with the `pyvistaqt.BackgroundPlotter()`? 

As for rendering in a background thread, this isn't (easily) possible and I actually ran into this (in a different context) here: https://github.com/Kitware/trame/discussions/173#discussioncomment-4605366. I'm blocked on MacOS due to the cocoa window needing to run from the main thread

**Reply by machur** [reply]
@banesullivan thanks for the quick response! 

Actually I started with https://docs.pyvista.org/api/plotting/plotting.html#plot-time-series-data and I tried to run that script from both command-line and PyCharm on Windows. I expected to see the visualization similar to the example on the website that would refresh itself. I didn't see any rendering at all though, the window froze and was unresponsive. It got closed automatically once the calculations were completed. 

I'm not sure if it's feasible considering the global interpreter lock and how Python handles threads. Right now I'm exploring options to deliver a stand-alone viewer application to the client with more friendly handling of processing. It's not acceptable for me that the whole application is unresponsive for couple of minutes while the meshes are being rendered.

**Reply by banesullivan** [reply]
We might be able to trigger the render event from a thread via `plotter.ren_win.InvokeEvent(vtk.vtkCommand.RenderEvent)`

Check out this example when using the trame-based viewer in Jupyter: cc @pyvista/developers @akaszynski 

```py
from threading import Thread
import time
import numpy as np
import pyvista as pv
import vtk

x = np.arange(-10, 10, 0.5)
y = np.arange(-10, 10, 0.5)
x, y = np.meshgrid(x, y)
r = np.sqrt(x**2 + y**2)
z = np.sin(r)

grid = pv.StructuredGrid(x, y, z)
grid['Height'] = z.ravel()

plotter = pv.Plotter()
plotter.add_mesh(
    grid,
    scalars="Height",
    lighting=False,
    show_edges=True,
    clim=[-1, 1],
)
plotter.show()
```

```py
def wiggle():
    while True:  # repeats forever
        nframe = 15
        for phase in np.linspace(0, 2 * np.pi, nframe + 1)[:nframe]:
            z = np.sin(r + phase)
            grid.points[:, -1] = z.ravel()
            grid['Height'] = z.ravel()
            plotter.ren_win.InvokeEvent(vtk.vtkCommand.RenderEvent)
            time.sleep(0.1)

thread = Thread(target=wiggle)
thread.start()
```


https://user-images.githubusercontent.com/22067021/218899816-0a7dba96-8d68-46f6-b2e9-dbfac1e9b344.mov

**Reply by banesullivan** [reply]
We should expose this event invocation on the Plotter API

**Reply by larsoner** [reply]
Maybe a new kwarg `async=False` (by default) where `plotter.render(async=True)` means "send this as an event rather than immediately rendering"?

**Reply by banesullivan** [reply]
Or what if we just automaticcaly did this anytime outside of the main thread via this check:

```py
import threading

if threading.current_thread() is threading.main_thread():
    ... # invoke event
else:
    ... # do render stuff
```

My concern with leaving it up to the user is that calling `render()` without the `async` argument would lead to a segfault and we likely don't want to shock users with that

**Reply by larsoner** [reply]
Sure if it can be automatic that's great!

**Reply by larsoner** [reply]
... we could make `async=None` use this automated behavior, leaving `async=True | False` for advanced users in case there's some issue with the automated one. But maybe YAGNI...

**Reply by machur** [reply]
Thank you guys for all the help! I've embedded your code into my viewer and it worked as intended. 

I like the concept of updating the underlying data and just sending the VTK event to render. My setup is different though since my scene contains hundreds of meshes/points/segments and I add them to the plotter one by one. I tried to run different variations of the following snippet (where self is the plotter - an implementation of QtInteractor):

```
def run_in_thread():
    print(f"Generating visualization...")

    self.add_mesh(pv.Sphere(radius=5), name="sphere1", render=False)
    self.add_mesh(pv.Sphere(radius=10), name="sphere2", render=False)
    self.add_mesh(pv.Sphere(radius=15), name="sphere3", render=False)

    print(f"Generating visualization completed")

thread = Thread(target=run_in_thread)
thread.start()
```

The joining of threads worked rarely and only for 2-3 objects with different names. When I try adding spheres in a loop or with sleeps between calls, it was really unstable. I kept getting the following errors:

```
2023-02-15 19:26:16.618 (   4.034s) [                ]vtkWin32OpenGLRenderWin:267    ERR| vtkWin32OpenGLRenderWindow (000002B4C1DDA300): wglMakeCurrent failed in MakeCurrent(), error: �숤ʴ
ERROR:root:wglMakeCurrent failed in MakeCurrent(), error: �숤ʴ
2023-02-15 19:26:16.619 (   4.035s) [                ]vtkWin32OpenGLRenderWin:267    ERR| vtkWin32OpenGLRenderWindow (000002B4C1DDA300): wglMakeCurrent failed in MakeCurrent(), error: 쌰왜ʴ
ERROR:root:wglMakeCurrent failed in MakeCurrent(), error: 쌰왜ʴ
2023-02-15 19:26:16.620 (   4.035s) [                ]vtkWin32OpenGLRenderWin:267    ERR| vtkWin32OpenGLRenderWindow (000002B4C1DDA300): wglMakeCurrent failed in MakeCurrent(), error: 쏐왜ʴ
ERROR:root:wglMakeCurrent failed in MakeCurrent(), error: 쏐왜ʴ
2023-02-15 19:26:16.620 (   4.036s) [                ]vtkOpenGLVertexArrayObj:278    ERR| vtkOpenGLVertexArrayObject (000002B4C65CC290): attempt to add attribute without a handleless buffer for attribute diffuseColor
ERROR:root:attempt to add attribute without a handleless buffer for attribute diffuseColor
2023-02-15 19:26:16.621 (   4.036s) [                ]vtkOpenGLVertexBufferOb:280    ERR| vtkOpenGLVertexBufferObjectGroup (000002B4C682ACC0): Error setting 'diffuseColor' in shader VAO.
ERROR:root:Error setting 'diffuseColor' in shader VAO.
2023-02-15 19:26:16.621 (   4.037s) [                ]vtkOpenGLVertexArrayObj:278    ERR| vtkOpenGLVertexArrayObject (000002B4C65CC290): attempt to add attribute without a handleless buffer for attribute vertexWC
ERROR:root:attempt to add attribute without a handleless buffer for attribute vertexWC
2023-02-15 19:26:16.622 (   4.037s) [                ]vtkOpenGLVertexBufferOb:280    ERR| vtkOpenGLVertexBufferObjectGroup (000002B4C682ACC0): Error setting 'vertexWC' in shader VAO.
ERROR:root:Error setting 'vertexWC' in shader VAO.
```

@banesullivan @larsoner Do you have any idea how to address this issue?
